addons/controllers/calico/calicoconfig_controller_test.go

```diff
@@ -0,0 +1,154 @@
+	cniv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cni/v1alpha1"
+)
+
+const (
+	waitTimeout     = time.Second * 15
+	pollingInterval = time.Second * 2
+)
+
+var _ = Describe("CalicoConfig Reconciler", func() {
+	var (
+		clusterName             string
+		clusterResourceFilePath string
+	)
+
+	JustBeforeEach(func() {
+		// create cluster resources
+		By("Creating a cluster and a CalicoConfig")
+		f, err := os.Open(clusterResourceFilePath)
+		Expect(err).ToNot(HaveOccurred())
+		defer f.Close()
+		err = testutil.CreateResources(f, cfg, dynamicClient)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Creating kubeconfig for cluster")
+		Expect(testutil.CreateKubeconfigSecret(cfg, clusterName, "default", k8sClient)).To(Succeed())
+	})
+
+	AfterEach(func() {
+		By("Deleting cluster and CalicoConfig")
+		f, err := os.Open(clusterResourceFilePath)
+		Expect(err).ToNot(HaveOccurred())
+		defer f.Close()
+		testutil.DeleteResources(f, cfg, dynamicClient, true)
+
+		By("Deleting kubeconfig for cluster")
+		key := client.ObjectKey{
+			Namespace: "default",
+			Name:      secret.Name(clusterName, secret.Kubeconfig),
+		}
+		s := &v1.Secret{}
+		Expect(k8sClient.Get(ctx, key, s)).To(Succeed())
+		Expect(k8sClient.Delete(ctx, s)).To(Succeed())
+	})
+
+	Context("reconcile CalicoConfig for management cluster", func() {
+
+		BeforeEach(func() {
+			clusterName = "test-cluster-1"
+			clusterResourceFilePath = "testcases/calico-test-1.yaml"
+		})
+
+		It("Should reconcile CalicoConfig and create data value secret on management cluster", func() {
+
+			key := client.ObjectKey{
+				Namespace: "default",
+				Name:      "test-cluster-1",
+			}
+
+			Eventually(func() bool {
+				cluster := &clusterapiv1beta1.Cluster{}
+				if err := k8sClient.Get(ctx, key, cluster); err != nil {
+					return false
+				}
+				return true
+			}, waitTimeout, pollingInterval).Should(BeTrue())
+
+			Eventually(func() bool {
+				config := &cniv1alpha1.CalicoConfig{}
+				if err := k8sClient.Get(ctx, key, config); err != nil {
+					return false
+				}
+
+				Expect(config.Spec.Namespace).Should(Equal("kube-system"))
+				Expect(config.Spec.InfraProvider).Should(Equal("vsphere"))
+				Expect(config.Spec.IPFamily).Should(Equal("ipv4"))
+				Expect(config.Spec.Calico.Config.VethMTU).Should(Equal(int64(0)))
+
+				return true
+			}, waitTimeout, pollingInterval).Should(BeTrue())
+
+			Eventually(func() bool {
+				secretKey := client.ObjectKey{
+					Namespace: "default",
+					Name:      fmt.Sprintf("%s-%s-data-values", clusterName, constants.CalicoAddonName),
+				}
+				secret := &v1.Secret{}
+				if err := k8sClient.Get(ctx, secretKey, secret); err != nil {
+					return false
+				}
+				Expect(secret.Type).Should(Equal(v1.SecretTypeOpaque))
+				secretData := string(secret.Data["values.yaml"])
+				Expect(strings.Contains(secretData, "IPFamily: ipv4")).Should(BeTrue())
+				return true
+			}, waitTimeout, pollingInterval).Should(BeTrue())
```

Comment:
Could you add the test case to check secretRef after successful reconciliation?

Reply:
Thanks, this time it worked. I guess the Deepcopy() I had  initially added was overwriting it undesirably